### ISSUE TITLE
Ranking Losses may be incorrectly implemented.

### ISSUE THREAD

**Issue by hello-fri-end** [mlpack/mlpack#3167]
I'm not an expert on Ranking Losses but I was reading up online and found out that they are mostly used for tasks like face verification using Siamese nets or Triplet nets.To use a Ranking Loss function we first extract features from two (or three) input data points and get an embedded representation for each of them. Then, we define a metric function to measure the similarity between those representations, for instance euclidian distance. Finally, we train the feature extractors to produce similar representations for both inputs, in case the inputs are similar, or distant representations for the two inputs, in case they are dissimilar
![pairwise_ranking_loss_faces](https://user-images.githubusercontent.com/43880587/157838232-0f9df92a-4afe-4cdf-9eb9-22f1a2d453c7.png)

Consider **Margin Ranking Loss**:
It's defined as:
![Screenshot from 2022-03-11 15-00-51](https://user-images.githubusercontent.com/43880587/157840421-11367c29-ebf9-4128-a3e2-ec24dd3dcb5a.png)
In mlpack, `x1` & `x2` are passed to the forward & backward function as one concatenated matrix, predictions, and `y` is the target matrix. In #3144 , I realized that the backward method is incorrect as it returns `dl / dy` instead of `dl / dx1` & `dl / dx2` .So, I [updated](https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/ann/loss_functions/margin_ranking_loss_impl.hpp#L73) the backward function to return the concatenated matrix `dl / dx1` & `dl / dx2` (this is what the backward function of Pytorch returns).

Consider **Cosine Embedding Loss**:
It's defined as: 
![Screenshot from 2022-03-11 14-21-23](https://user-images.githubusercontent.com/43880587/157838367-470a396b-e2fb-4ba4-bf01-144765e60f34.png)

In the current implementation, `x1` is the predictions matrix, `x2` is the target and the backward function returns `dl/dx1`. But just like in Margin Ranking Loss, I feel that  predictions matrix should be a concatenation of `x1` and `x2` ( anchor & positive embedding or anchor & negative embedding from a siamese network for example) and the backward function should return the concatenated gradient `dl/dx1` & `dl/dx2` ( Pytorch also return both the gradients). 

Also, [`batchSize`](https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/ann/loss_functions/cosine_embedding_loss_impl.hpp#L77) is incorrectly calculated in the current implementation, it assumes that each row represents a sample while mlpack assumes each column is one sample.

Finally, **Triplet Margin Loss**:
![Screenshot from 2022-03-11 15-15-07](https://user-images.githubusercontent.com/43880587/157842762-70ec4326-0653-46b1-9697-f0bdfdb5dfbc.png)

It's used in triplet networks (three CNNs with shared weights). Currently, `a` & `p` are passed to the forward & backward function as one concatenated matrix, predictions & `n` is the target. And the backward function only returns gradient with respect to `p`. I think here the predictions matrix should be a concatenation of `a`, `p` & `n` & the backward method should return gradient with respect to all three ( Pytorch also returns the same)..

But as I said, I'm not an expert on Ranking Losses. Maybe I'm missing something. Looking forward to some discussion on this.,
